### PR TITLE
Batched RPA - Optimize KV reshape in prepare_inputs

### PR DESCRIPTION
**Pull request by pritha90**
# Description

KV reshape in batched RPA kernel are taking longer than RPAv3, primarily because of copies stemming from  with_layout_constraint. This is primarily required when physical memory layout of K/V are {2,1,0} -> observed in kv_heads >=4 case of Gemma4.
Added 2 paths to reshape depending on num_kv_heads

# Tests

1. Gemma4 compiles and server is up.
Gemma4 has 2 variants of the K & V vectors:

a. f8e4m3fn[16,4,256]{2,1,0:T(4,128)(4,1)} => This vector needs the layout constraint to confirm with the generic padding norms.

b. f8e4m3fn[16,1,512]{2,0,1:T(8,128)(4,1)}  

USE_BATCHED_RPA_KERNEL=1 MODEL_IMPL_TYPE=flax_nnx vllm \
    serve google/Gemma-4-26B-A4B-it \
    --max-model-len=16384 \
    --max-num-seqs=320 \
    --tensor-parallel-size 2 \
    --max-num-batched-tokens 4096 \
    --no-enable-prefix-caching \
    --kv-cache-dtype=fp8

2. Qwen3.5-397B-A17B-FP8 with Batched RPA picks the faster reshape path

8K/1K concurrency 512 ATTN_DP,EP8 

============ Serving Benchmark Result ============
Successful requests:                     640       
Benchmark duration (s):                  281.53    
Total input tokens:                      4727544   
Total generated tokens:                  589927    
Request throughput (req/s):              2.27      
Output token throughput (tok/s):         2095.42   
Total Token throughput (tok/s):          18887.64  

Throughput gain 18887 / 18136.78128 (basline)

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by lk-chen** [comment]
could you verify the command from https://github.com/vllm-project/tpu-inference/pull/2506 still works?

File: tpu_inference/kernels/experimental/batched_rpa/wrapper.py [hunk 1]
```diff
@@ -91,30 +90,18 @@ def prepare_inputs(
         f"{num_kv_heads_x2_aligned}")
     num_kv_heads_aligned = num_kv_heads_x2_aligned // 2
 
-    # `jnp.stack` introduces the K/V axis as a fresh axis, keeping
-    # the data 4D `[T, H_kv, 2, D]` through the pad. The final reshape
-    # `[T, H_kv, 2, D] -> [T, A/P, P, D]` becomes a pure factor merge of
```
Comment (pritha90) [inline tpu_inference/kernels/experimental/batched_rpa/wrapper.py:96]: Reshape+concat  instead of stack

T, KV, D

**Comment by pritha90** [comment]
> could you verify the command from #2506 still works?

https://paste.googleplex.com/6250953515859968 -> works

File: tpu_inference/kernels/experimental/batched_rpa/wrapper.py [hunk 2]
```diff
@@ -86,40 +86,64 @@ def prepare_inputs(
...
-    # `with_layout_constraint` on `kv_padded` pins the pad output
-    # to major-to-minor (0, 1, 2, 3) — i.e., T majormost, head_dim minor.
-    # Without it XLA's auto-layout placed the pad in SMEM with a
-    # transposed dim order, requiring a layout-changing copy to reach the
-    # kernel's required HBM layout — and the LLO emitter cannot lower
-    # that simultaneous axis-swap + tile-resize + memory-space migration.
-    kv_padded = with_layout_constraint(
-        jnp.pad(
-            kv_stacked,
+    if actual_num_kv_heads > 2:
```
Comment (a1yssan13) [inline tpu_inference/kernels/experimental/batched_rpa/wrapper.py]: Hmm, why is this > 2 but the comment says heads >= 4? Have we done any testing on what head sizes trigger this copy

**Review by a1yssan13 — APPROVED** [review]

File: tpu_inference/kernels/experimental/batched_rpa/wrapper.py (hunk 2 above)
Comment (pritha90) [reply tpu_inference/kernels/experimental/batched_rpa/wrapper.py]: Tested 1, 2, 4 so far. And the 4 was only tested on gemma4. 1, 2 on qwen 3.5
KV heads are usually multiples of 2 , hence made the check on > 2

Comment (kyuyeunk) [reply tpu_inference/kernels/experimental/batched_rpa/wrapper.py]: is it possible to generalize? maybe the issue only happens when `actual_num_kv_heads==packing`? for instance, do you see the error when using bf16 which has `packing=2`?

Comment (pritha90) [reply tpu_inference/kernels/experimental/batched_rpa/wrapper.py]: bf16[16,4,256]{2,1,0:T(4,128)(2,1)}  => issue seen on bf16 too
vs
bf16[16,1,512]{2,0,1:T(8,128)(2,1)} for kv_head=1

Comment (pritha90) [reply tpu_inference/kernels/experimental/batched_rpa/wrapper.py]: Will add the hack in for now, until we get to the bottom of gemma4 layout

**Review by kyuyeunk — APPROVED** [review]
